### ISSUE TITLE
CleanUp not working as expected when previous tests fail 

### ISSUE THREAD

**Issue by shoutcool** [database-rider/database-rider#74]
i have 2 datasets:

**document_singleEntry.yml**
```
DOCTABLE:
  - IDDOCUMENT: 1
    TEXT: This is my test
```

**document_emptyTable.yml**
```
DOCTABLE:
```

And i have 2 Unit Test using the above datasets:
```
@Test
@DataSet(value = "document_singleEntry.yml", transactional = false, cleanBefore = true)
public void readAll_success_returnsElement() {
	DocumentDao testee = new DocumentDao(emProvider.getEm());
	List<Document> results = testee.readAll();
	assertThat(results , hasSize(1));

	Document result = results .get(0);

	assertEquals(Integer.valueOf(1), result.getId());
	assertEquals("This is my test", result.getText());
}
```

```
@Test
@DataSet(value = "document_emptyTable.yml", transactional = true, cleanBefore = true)
@ExpectedDataSet(value = "document_singleEntry.yml")
public void insert_success_returnsElement() {
	DocumentDao testee = new DocumentDao(emProvider.getEm());

	Document doc = new Document();
        doc.setId(1);
        doc.setText("This is my test");

        testee.merge(doc)
}
```

If i execute both tests, everything is fine. But if i break the second one by setting:
`doc.setText("This is my FAILING test");`
then both tests are failing! 

So the tests are not independent. The order of execution was:

1. insert_success_returnsElement
2. readAll_success_returnsElement

First Test is failing because of the wrong text i provided --> correct
Second Test is failing because the Text "This is my FAILING test" is still in the database.

But why? if you take a look at the readAll_success_returnsElement Test, then you see that i've set the cleanBefore Attribute AND i specified the document_singleEntry.yml dataset for initial data. But nevertheless the wrong data from the insert Test is still in the Database.

This is not the expected behaviour IMHO.

I tested against a H2 In-Memory Database. 

Thank you for your help.

**Comment by rmpestano** [comment]
Hi @shoutcool,

can you try cleaning entity manager cache?  

```
@Before
public void clearCache() {
   emProvider.getEm().clear();
}

@Test
@DataSet(value = "document_singleEntry.yml", transactional = false, cleanBefore = true)
public void readAll_success_returnsElement() {
	DocumentDao testee = new DocumentDao(emProvider.getEm());
	List<Document> results = testee.readAll();
	assertThat(results , hasSize(1));

	Document result = results .get(0);

	assertEquals(Integer.valueOf(1), result.getId());
	assertEquals("This is my test", result.getText());
}

@Test
@DataSet(value = "document_emptyTable.yml", transactional = true, cleanBefore = true)
@ExpectedDataSet(value = "document_singleEntry.yml")
public void insert_success_returnsElement() {
	DocumentDao testee = new DocumentDao(emProvider.getEm());

	Document doc = new Document();
        doc.setId(1);
        doc.setText("This is my test");

        testee.merge(doc)
}

```

**Comment by rmpestano** [comment]
I didn't had time to test but I think I've found the problem. 

If you are using EntityManagerProvider as a JUnit Rule [like here](https://github.com/database-rider/database-rider/blob/master/rider-core/src/test/java/com/github/database/rider/core/CrudIt.java#L28), it should clear entity manager [cache here](https://github.com/database-rider/database-rider/blob/2026adc92151e72b7061ad87cfc4cd6f7e042387/rider-core/src/main/java/com/github/database/rider/core/util/EntityManagerProvider.java#L204) but We  should do it in a [finally block](url) otherwise the clear will not be reached when a exception raised (your case I think).

**Comment by shoutcool** [comment]
Thank you for your hints 👍 

I tested several ways according to your comments:

**Normal Configuration was like this:**

```
@Rule
public EntityManagerProvider emProvider = EntityManagerProvider.instance("portal-test");

@Rule
public DBUnitRule dbUnitRule = DBUnitRule.instance(emProvider.connection());
```

Result: Tests are **not** independent (according to my description above)

**Your hint no. 1:**

```
@Rule
public EntityManagerProvider emProvider = EntityManagerProvider.instance("portal-test");

@Rule
public DBUnitRule dbUnitRule = DBUnitRule.instance(emProvider.connection());

@Before
public void clearCache() {
	EntityManagerProvider.clear();
}
```

Result: **JACKPOT: Everything works as expected --> THANK YOU!**

**Your hint no. 2**

```
public EntityManagerProvider emProvider = EntityManagerProvider.instance("portal-test");

@Rule
public TestRule theRule = RuleChain.outerRule(emProvider).around(DBUnitRule.instance(emProvider.connection()));
```

Result: Tests are **not** independent 

So i will stick with clearing the EntityManager Cache, seems to work fine. 👍 Mabye this should be clarified in the docs, or even better not be necessary. Thank you anyway :)

**Comment by rmpestano** [comment]
Agreed, It should be automatic. Thanks for reporting.

**Comment by rmpestano** [comment]
Hi @shoutcool ,

Can you try with rider `1.2.8-SNAPSHOT`? the emProvider.clear() should not be needed anymore.

You'll need to declare snapshots repository:

```
<repositories>
    <repository>
        <snapshots/>
        <id>snapshots</id>
        <name>libs-snapshot</name>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
    </repository>
</repositories>
```

**Comment by shoutcool** [comment]
Version 1.2.8-SNAPSHOT behaves as expected. Works without cleaning the EM's cache manually 👍

**Comment by rmpestano** [comment]
Just released v1.2.8 containing this fix.
